File: Data_Structures/majority_voting.cpp

// O((n+q) log n)
struct MajorityDS {

struct Node {
    int cand; // possible majority candidate
    int cnt;  // net dominance after cacllations
};

static Node cmb(Node a, Node b) {
    if (a.cnt == 0)  return b;
    if (b.cnt == 0)  return a;
    if (a.cand == b.cand) return {a.cand, a.cnt + b.cnt};
    if (a.cnt > b.cnt) return {a.cand, a.cnt - b.cnt};
    if (b.cnt > a.cnt) return { b.cand, b.cnt - a.cnt};
    return {0, 0};
}

static vector<int> a;

struct SegTree {
    int n;
    vector<Node> t;

    void init(int _n) {
        n = _n;
        t.resize(4*n);
    }

    void build(const vi& a, int v, int tl, int tr) {
        if (tl == tr) t[v] = Node{a[tl], 1};
        else {
            int tm = tl+(tr-tl)/2;
            build(a, v*2, tl, tm);
            build(a, v*2+1, tm+1, tr);
            t[v] = cmb(t[v*2], t[v*2+1]);
        }
    }

    void update(int i, int val) {
        a[i] = val;
        update(1, 0, n-1, i, val);
    }

    void update(int v, int tl, int tr, int posi, int val) {
        if (tl == tr) t[v] = Node{val, 1};
        else {
            int tm = tl+(tr-tl)/2;
            if (posi <= tm) {
                update(v*2, tl, tm, posi, val);
            } else {
                update(v*2+1, tm+1, tr, posi, val);
            }
            t[v] = cmb(t[v*2], t[v*2+1]);
        }
    }

    Node query(int v, int tl, int tr, int l, int r) {
        if (l > r) return {0, 0};
        if (l == tl && r == tr) return t[v];
        int tm = tl+(tr-tl)/2;
        return cmb(
            query(v*2, tl, tm, l, min(r, tm)),
            query(v*2+1, tm+1, tr, max(l, tm+1), r)
        );
    }
};

template<typename T>
struct FenwickTree {
    int n;
    vector<T> fw;

    FenwickTree(int _n = 0) : n(_n) { fw.assign(n+1, 0); }

    static int lsb(int i) { return i & -i; }

    void build(const vector<T>& a) {
        n = sz(a) + 1;
        for (int i = 1; i < n; i++) {
            fw[i] += a[i-1];
            if (i + lsb(i) <= sz(a)) {
                fw[i + lsb(i)] += fw[i];
            }
        }
    }

    void update(int i, T val) {
        for (++i; i < n; i += lsb(i))
            fw[i] += val;
    }

    T sum(int i) {
        T res = 0;
        for (++i; i > 0; i -= lsb(i)) res += fw[i];
        return res;
    }
};

struct FenwickTree01 {
    vector<u64> a;       // bit blocks
    #undef ft
    FenwickTree<int> ft; // counts of ones per block

    FenwickTree01(int n = 0) {
        int m = (n + 63) / 64;
        a.resize(m);
        ft = FenwickTree<int>(m);
    }

    void set(int i, bool f) {
        int j = i % 64; // bit inside block
        i /= 64;        // block index
        int x = (a[i] >> j) & 1;    // current bit
        if (x && !f) {
            a[i] ^= 1ULL << j;
            ft.update(i, -1);
        } else if (!x && f) {
            a[i] |= 1ULL << j;
            ft.update(i, 1);
        }
    }

    // prefix sum query [0, i)
    int sum(int i) {
        int block = i / 64;
        int bit = i % 64;

        int s = 0;
        if (block > 0)
            s = ft.sum(block - 1); // full blocks [0, block)
        if (bit && block < sz(a)) s += __builtin_popcountll(a[block] & ((1ULL << bit) - 1));
        return s;
    }

    int rangeSum(int l, int r) {
        return sum(r) - sum(l);
    }
};


signed main(void) {
    int n, q; re(n, q);
    a.resize(n);
    for (int i = 0; i < n; i++) cin >> a[i];

    vector<array<int, 3>> queries(q);
    vector<int> vals = a;
    for (int i = 0; i < q; i++) {
        cin >> queries[i][0] >> queries[i][1] >> queries[i][2];
        if (queries[i][0] == 0) vals.push_back(queries[i][2]);
    }

    // coordinate compression
    sort(all(vals));
    vals.erase(unique(all(vals)), vals.end());
    auto id = [&vals](int x) {
        return lower_bound(all(vals), x) - vals.begin();
    };

    for (int i = 0; i < n; i++) a[i] = id(a[i]);

    for (auto &[t, p, x] : queries)
        if (t == 0) x = id(x);

    /* static positions list
       xs[v] = all indices where value v can ever appear */
    int m = sz(vals);
    vector<vector<int>> xs(m);
    // initial array
    for (int i = 0; i < n; i++)
        xs[a[i]].push_back(i);

    // future updates
    for (auto [t, p, x] : queries) {
        if (t == 0) xs[x].push_back(p);
    }

    // clean it up
    for (int v = 0; v < m; v++) {
        sort(xs[v].begin(), xs[v].end());
        xs[v].erase(unique(xs[v].begin(), xs[v].end()), xs[v].end());
    }

    // Fenwick01 per value
    vector<FenwickTree01> occ(m);
    for (int v = 0; v < m; v++)
        occ[v] = FenwickTree01(xs[v].size());

    for (int i = 0; i < n; i++) {
        int v = a[i];
        int idx = lower_bound(all(xs[v]), i) - xs[v].begin();
        occ[v].set(idx, 1);
    }

    // --- segment tree ---
    SegTree st;
    st.init(n);
    st.build(a, 1, 0, n-1);

    // --- process queries ---
    for (auto [t, x, y] : queries) {
        if (t == 0) { // update
            int old = a[x];
            if (old != y) {
                int idx_old = lower_bound(all(xs[old]), x) - xs[old].begin();
                occ[old].set(idx_old, 0);

                int idx_new = lower_bound(all(xs[y]), x) - xs[y].begin();
                occ[y].set(idx_new, 1);

                a[x] = y;
                st.update(x, y);
            }
        } else { // query [x, y)
            int l = x, r = y-1;
            Node res = st.query(1, 0, n-1, l, r);

            if (res.cnt == 0) {
                cout << -1 << "\n";
                continue;
            }

            int v = res.cand;
            int L = lower_bound(all(xs[v]), x) - xs[v].begin();
            int R = lower_bound(all(xs[v]), y) - xs[v].begin();
            int freq = occ[v].rangeSum(L, R);

            if (freq * 2 > (y - x))
                cout << vals[v] << "\n";
            else
                cout << -1 << "\n";
        }
    }

    return 0;
}
}; // struct MajorityDS


// SLOW VERSION! FOR LEARNING PURPOSES! vector.erase() is the killer here, need something better
// For example: Fenwick over Fenwick01
namespace slow {
struct Node {
    int cand; // possible majority candidate
    int cnt;  // net dominance after cacllations
};

Node cmb(Node a, Node b) {
    if (a.cnt == 0)  return b;
    if (b.cnt == 0)  return a;
    if (a.cand == b.cand) return {a.cand, a.cnt + b.cnt};
    if (a.cnt > b.cnt) return {a.cand, a.cnt - b.cnt};
    if (b.cnt > a.cnt) return { b.cand, b.cnt - a.cnt};
    return {0, 0};
}

vector<int> a = {1, 5, 5, 2, 4, 3, 2};
unordered_map<int, vector<int>> pos;

struct SegTree {
    int n;
    vector<Node> t;

    void init(int _n) {
        n = _n;
        t.resize(4*n);
    }

    void build(const vi& a, int v, int tl, int tr) {
        if (tl == tr) t[v] = Node{a[tl], 1};
        else {
            int tm = tl+(tr-tl)/2;
            build(a, v*2, tl, tm);
            build(a, v*2+1, tm+1, tr);
            t[v] = cmb(t[v*2], t[v*2+1]);
        }
    }

    void update(int i, int val) {
        pos[a[i]].erase(remove(pos[a[i]].begin(), pos[a[i]].end(), i), pos[a[i]].end());
        pos[val].push_back(i);
        a[i] = val;
        update(1, 0, n-1, i, val);
    }

    void update(int v, int tl, int tr, int posi, int val) {
        if (tl == tr) t[v] = Node{val, 1};
        else {
            int tm = tl+(tr-tl)/2;
            if (posi <= tm) {
                update(v*2, tl, tm, posi, val);
            } else {
                update(v*2+1, tm+1, tr, posi, val);
            }
            t[v] = cmb(t[v*2], t[v*2+1]);
        }
    }

    Node query(int v, int tl, int tr, int l, int r) {
        if (l > r) return {0, 0};
        if (l == tl && r == tr) return t[v];
        int tm = tl+(tr-tl)/2;
        return cmb(
            query(v*2, tl, tm, l, min(r, tm)),
            query(v*2+1, tm+1, tr, max(l, tm+1), r)
        );
    }

};


int count(int v, int l, int r) {
    auto &vec = pos[v];
    return lower_bound(vec.begin(), vec.end(), r)
        - lower_bound(vec.begin(), vec.end(), l);
}


signed main(void) {
    for (int i = 0; i < sz(a); i++) {
        pos[a[i]].push_back(i);
    }
    int n = sz(a);
    SegTree t;
    t.init(n);
    t.build(a, 1, 0, n-1);

    int l = 1, r = 4;
    r--;
    auto res = t.query(1, 0, n-1, l, r);
    if (res.cnt == 0) print(-1);
    else {
        int freq = count(res.cand, l, r);
        if (freq * 2 > r-l) print(res.cand);
        else print(-1);
    }

    return 0;
}
} // namespace slow

File: datastructures/gol/majority_voting.go
package gol

import "sort"

// O((n+q) log n)
type MajorityDS struct {
	a   []int
	pos map[int][]int
	st  majoritySegTree
}

type majorityNode struct {
	cand int // possible majority candidate
	cnt  int // net dominance after cancellations
}

func combineMajority(a, b majorityNode) majorityNode {
	if a.cnt == 0 {
		return b
	}
	if b.cnt == 0 {
		return a
	}
	if a.cand == b.cand {
		return majorityNode{cand: a.cand, cnt: a.cnt + b.cnt}
	}
	if a.cnt > b.cnt {
		return majorityNode{cand: a.cand, cnt: a.cnt - b.cnt}
	}
	if b.cnt > a.cnt {
		return majorityNode{cand: b.cand, cnt: b.cnt - a.cnt}
	}
	return majorityNode{}
}

type majoritySegTree struct {
	n int
	t []majorityNode
}

func (s *majoritySegTree) init(n int) {
	s.n = n
	s.t = make([]majorityNode, 4*n)
}

func (s *majoritySegTree) build(a []int, v, tl, tr int) {
	if tl == tr {
		s.t[v] = majorityNode{cand: a[tl], cnt: 1}
	} else {
		tm := tl + (tr-tl)/2
		s.build(a, v*2, tl, tm)
		s.build(a, v*2+1, tm+1, tr)
		s.t[v] = combineMajority(s.t[v*2], s.t[v*2+1])
	}
}

func (s *majoritySegTree) update(v, tl, tr, posi, val int) {
	if tl == tr {
		s.t[v] = majorityNode{cand: val, cnt: 1}
	} else {
		tm := tl + (tr-tl)/2
		if posi <= tm {
			s.update(v*2, tl, tm, posi, val)
		} else {
			s.update(v*2+1, tm+1, tr, posi, val)
		}
		s.t[v] = combineMajority(s.t[v*2], s.t[v*2+1])
	}
}

func (s *majoritySegTree) query(v, tl, tr, l, r int) majorityNode {
	if l > r {
		return majorityNode{}
	}
	if l == tl && r == tr {
		return s.t[v]
	}
	tm := tl + (tr-tl)/2
	return combineMajority(
		s.query(v*2, tl, tm, l, minInt(r, tm)),
		s.query(v*2+1, tm+1, tr, maxInt(l, tm+1), r),
	)
}

func NewMajorityDS(a []int) *MajorityDS {
	ds := &MajorityDS{
		a:   append([]int(nil), a...),
		pos: make(map[int][]int),
	}
	for i, v := range ds.a {
		ds.pos[v] = append(ds.pos[v], i)
	}
	ds.st.init(len(ds.a))
	if len(ds.a) > 0 {
		ds.st.build(ds.a, 1, 0, len(ds.a)-1)
	}
	return ds
}

func (ds *MajorityDS) Update(i, val int) {
	old := ds.a[i]
	if old == val {
		return
	}

	oldVec := ds.pos[old]
	oldPos := sort.SearchInts(oldVec, i)
	if oldPos < len(oldVec) && oldVec[oldPos] == i {
		oldVec = append(oldVec[:oldPos], oldVec[oldPos+1:]...)
	}
	if len(oldVec) == 0 {
		delete(ds.pos, old)
	} else {
		ds.pos[old] = oldVec
	}

	newVec := ds.pos[val]
	insertPos := sort.SearchInts(newVec, i)
	newVec = append(newVec, 0)
	copy(newVec[insertPos+1:], newVec[insertPos:])
	newVec[insertPos] = i
	ds.pos[val] = newVec

	ds.a[i] = val
	ds.st.update(1, 0, len(ds.a)-1, i, val)
}

// Query returns the majority element on [l, r), if it exists.
func (ds *MajorityDS) Query(l, r int) (int, bool) {
	if !(0 <= l && l <= r && r <= len(ds.a)) {
		panic("Query: invalid range")
	}
	if l == r {
		return 0, false
	}
	res := ds.st.query(1, 0, len(ds.a)-1, l, r-1)
	if res.cnt == 0 {
		return 0, false
	}

	v := res.cand
	freq := ds.count(v, l, r)
	if freq*2 > (r - l) {
		return v, true
	}
	return 0, false
}

func (ds *MajorityDS) count(v, l, r int) int {
	vec := ds.pos[v]
	return sort.SearchInts(vec, r) - sort.SearchInts(vec, l)
}

// SLOW VERSION! FOR LEARNING PURPOSES! vector.erase() is the killer here, need something better
// For example: Fenwick over Fenwick01
type SlowMajorityDS struct {
	a   []int
	pos map[int][]int
	st  majoritySegTree
}

func NewSlowMajorityDS(a []int) *SlowMajorityDS {
	ds := &SlowMajorityDS{
		a:   append([]int(nil), a...),
		pos: make(map[int][]int),
	}
	for i, v := range ds.a {
		ds.pos[v] = append(ds.pos[v], i)
	}
	ds.st.init(len(ds.a))
	if len(ds.a) > 0 {
		ds.st.build(ds.a, 1, 0, len(ds.a)-1)
	}
	return ds
}

func (ds *SlowMajorityDS) Update(i, val int) {
	old := ds.a[i]
	oldVec := ds.pos[old]
	for j, x := range oldVec {
		if x == i {
			ds.pos[old] = append(oldVec[:j], oldVec[j+1:]...)
			break
		}
	}
	ds.pos[val] = append(ds.pos[val], i)
	sort.Ints(ds.pos[val])
	ds.a[i] = val
	ds.st.update(1, 0, len(ds.a)-1, i, val)
}

func (ds *SlowMajorityDS) Count(v, l, r int) int {
	vec := ds.pos[v]
	return sort.SearchInts(vec, r) - sort.SearchInts(vec, l)
}

func (ds *SlowMajorityDS) Query(l, r int) (int, bool) {
	if l >= r {
		return 0, false
	}
	res := ds.st.query(1, 0, len(ds.a)-1, l, r-1)
	if res.cnt == 0 {
		return 0, false
	}
	freq := ds.Count(res.cand, l, r)
	if freq*2 > (r - l) {
		return res.cand, true
	}
	return 0, false
}
